sound/soc/sof/intel/cnl.c

```diff
@@ -244,108 +244,38 @@ void cnl_ipc_dump(struct snd_sof_dev *sdev)
 }
 
 /* cannonlake ops */
-struct snd_sof_dsp_ops sof_cnl_ops = {
-	/* probe/remove/shutdown */
-	.probe		= hda_dsp_probe,
-	.remove		= hda_dsp_remove,
-	.shutdown	= hda_dsp_shutdown,
-
-	/* Register IO */
-	.write		= sof_io_write,
-	.read		= sof_io_read,
-	.write64	= sof_io_write64,
-	.read64		= sof_io_read64,
+struct snd_sof_dsp_ops sof_cnl_ops;
+EXPORT_SYMBOL_NS(sof_cnl_ops, SND_SOC_SOF_INTEL_HDA_COMMON);
 
-	/* Block IO */
-	.block_read	= sof_block_read,
-	.block_write	= sof_block_write,
+void sof_cnl_ops_init(struct snd_sof_dev *sdev)
+{
+	/* common defaults */
+	memcpy(&sof_cnl_ops, &sof_hda_common_ops, sizeof(struct snd_sof_dsp_ops));
 
-	/* Mailbox IO */
-	.mailbox_read	= sof_mailbox_read,
-	.mailbox_write	= sof_mailbox_write,
+	/* probe/remove/shutdown */
+	sof_cnl_ops.shutdown	= hda_dsp_shutdown;
 
 	/* doorbell */
-	.irq_thread	= cnl_ipc_irq_thread,
+	sof_cnl_ops.irq_thread	= cnl_ipc_irq_thread;
```

Comment:
and so how would we handle the change of this op for IPC3 vs IPC4? Wouldnt it be better so have a struct snd_sof_dsp_ipc_dependent_ops ipc_dep_ops[SOF_IPC_COUNT] inside snd_sof_dsp_ops and set to match the IPC version? and then the init can basically simply do sof_cnl_ops.ipc_dep_ops = cnl_ipc_dep_ops[pdata->ipc_type;

or we would need 2 init functions per IPC type.

Reply:
well, this is what I had before but you suggested to use code to set the ops as needed.

so I was thinking of 

````c
if (ipc_type == SOF_IPC)
    sof_cnl_ops.irq_thread = cnl_ipc_irq_thread;
else
    sof_cnl_ops.irq_thread = cnl_ipc_irq_thread_ipc4;
````

Reply:
but we can still have the ipc-dependent ops per IPC type and set it here. In fact I think we dont even need the ops_init, if we split up snd_sof_dsp_ops into snd_sof_ipc_independent_dsp_ops and snd_sof_ipc_dependent_ops. Then, the ipc_independent ops can be const and set like we do now and the ipc_dependent_ops can be set during probe() based on pdata->ipc_type

Reply:
I don't see how I can keep a 'const' qualifier and reduce the differences between platforms as I did in this PR.

the 'const' forces us to duplicate N-1 fields when only 1 changes, with 4 platforms already it's not really efficient any longer.

Reply:
I meant to keep the const only for the IPC-independent bits. Let me cook up a patch quickly

Reply:
@plbossart this is what I had in mind on top what you have already. https://github.com/ranj063/linux/commit/f7921a00ee2296d1e1ce6e1a8588df48091c1cdb

I only separated out the irq_thread op for now just to show the idea.

Reply:
@ranj063 the const will not work even for the IPC-independent bits. When we add SKL, we will need a different loader even for the same IPC.